numba/np/ufunc/dufunc.py

```diff
@@ -231,6 +242,260 @@ def _compile_for_argtys(self, argtys, return_type=None):
+                flat_idx, i, j, len_idx = 0, 0, 0, len(idx)
+                while i < len_idx:
+                    if i != axis:
+                        flat_idx += strides[j] * idx[i]
+                        j += 1
+                    i += 1
+                flat_idx //= itemsize
+                return flat_idx
+
+            @register_jitable
+            def find_min(tup):
+                idx, e = 0, tup[0]
+                for i in range(len(tup)):
+                    if tup[i] < e:
+                        idx, e = i, tup[i]
+                return idx, e
+
+            def impl_1d(ufunc, array, axis=0, dtype=None, initial=None):
+                if init_none and id_none:
+                    r = array[0]
+                elif init_none:
+                    r = identity
+                else:
+                    r = initial
+
+                sz = array.shape[0]
+                for i in range(sz):
+                    r = ufunc(r, array[i])
+                return r
+
+            def impl_nd_axis_int(ufunc,
+                                 array,
+                                 axis=0,
+                                 dtype=None,
+                                 initial=None):
+                if axis is None:
+                    raise ValueError("'axis' must be specified")
+
+                if axis < 0 or axis >= array.ndim:
+                    raise ValueError("Invalid axis")
+
+                # create result array
+                shape = tuple_slice(array.shape, axis)
+
+                if initial is None and identity is None:
+                    r = np.empty(shape, dtype=nb_dtype)
+                    for idx, _ in np.ndenumerate(r):
+                        # shape[0:axis] + 0 + shape[axis:]
+                        result_idx = tuple_slice_append(idx, axis, 0)
+                        r[idx] = array[result_idx]
+                elif initial is None and identity is not None:
+                    # Checkig if identity is not none is redundant but required
+                    # compile this block
+                    r = np.full(shape, fill_value=identity, dtype=nb_dtype)
+                else:
+                    r = np.full(shape, fill_value=initial, dtype=nb_dtype)
+
+                # One approach to implement reduce is to remove the axis index
+                # from the indexing tuple returned by "np.ndenumerate". For
+                # instance, if idx = (X, Y, Z) and axis=1, the result index
+                # is (X, Y).
+                # Another way is to compute the result index using strides,
+                # which is faster than manipulating tuples.
+                view = r.ravel()
+                for idx, val in np.ndenumerate(array):
+                    flat_pos = compute_flat_idx(r.strides, r.itemsize, idx,
+                                                axis)
+                    lhs, rhs = view[flat_pos], val
+                    view[flat_pos] = ufunc(lhs, rhs)
+                return r
```

Comment:
This works, but I think the re-computation of flat_pos from basics at each iteration is going to be a bit slow. I would think you could avoid `nd.enumerate` and have a loop that only increments the flat_pos.

I wonder how the non-reduce ufuncs do this indexing, since they too can accept an axis argument.